### DISCUSSION TITLE
Sharing piece of state into a optional child of a child state

### DISCUSSION THREAD

**Discussion by mapedd** [pointfreeco/swift-composable-architecture#1402] [Q&A]
Hello. I'm trying to figure what's the best way to model something like this

<img width="1079" alt="Screenshot 2022-09-20 at 14 41 08" src="https://user-images.githubusercontent.com/422167/191260467-90a33eb6-2e0d-45f9-a963-c6860d7bf802.png">

I have a `MainState` that holds into an `records` array of `Records`. 
`MainView` is showing and manipulating it.
`MainState` has a child called `SummaryState` that does not need to know anything about the `records`. But then `SummaryState` has an optional piece of state  called `StatisticsState`, that wants to show `records` in various way by filtering/sorting it. 

I think using `Environment` to read/write to `records` array would remove a lot of the benefits of using TCA.
Adding `records` to `SummaryState` just so that it's then passed to `StatisticsState`  seems like a heavy solution that will be difficult to maintain. 

Any suggestions?

**Comment by Alex293** [comment]
You can move the navigation to the stats feature up in the main feature (stat feature become a child of main feature) or your summary feature could take a viewBuilder closure so it can build and navigate to stats without knowing how to build it. Then when you build the summary view you pass a view builder.

One thing that I like to keep in mind is that your features (state, action, reducers ...) structure can vastly differ from the navigation hierarchy. Also as there is only one store you can pass a non scoped store to subviews, just be careful to scope it where you use it in viewStores.

**Reply by mapedd** [reply]
Those are some good suggestions @Alex293 , thanks 👍

**Reply by mapedd** [reply]
For the first suggestion, I think the problem would be, that action that is sent from the `SummaryView` ( so should be of type `SummaryAction`) needs to then referenced in the `MainView` in the sheet presentation. I'd expect to use only action from the single domain in the view for a given state.

```
MainView.swift


  .sheet(
        isPresented: viewStore.binding(
          get: \.showStatistics,
          send: { $0 ? SummaryAction.showStatistics : SummaryAction.hideStatistics }
        )
      )
```

**Reply by Alex293** [reply]
If main is the entry point it can get access to all the actions which you can wrap like : 
```
.sheet(
        isPresented: viewStore.binding(
          get: \.showStatistics,
          send: { .summary($0 ? SummaryAction.showStatistics) : SummaryAction.hideStatistics) }
        )
      )
```
but really if main dispatches and interprets the actions for navigating those action should go into main feature.

I would do something like: 
```
enum SummaryAction {
  ...
  case delegate(DelegateAction)

  enum DelegateAction {
    case findAGoodnameToPushStatistics
  }
}
```
and you return .none in Summary reducer but in your main reducer you handle this action.

The idea behind having a separate delegate is to make clear that you don't have to do anything in the reducer but that those need to be handled by the user of the feature.
